Metagenomics Remove Human Reads (#93)
nasa/GeneLab_Data_Processing: Metagenomics/Remove_human_reads_from_raw_data/Workflow_Documentation/NF_MGRemoveHumanReads-A/workflow_code/README.md

## Utilizing the workflow

1. [Install conda, mamba, and `genelab-utils` package](#1-install-conda-mamba-and-genelab-utils-package) 
2. [Install NextFlow and Singularity](#2-install-NextFlow-Singularity)
   2a. [Install Nextflow](#2a-install-nextflow)  
   2b. [Install Singularity](#2b-install-singularity)
3. [Download the workflow template files](#3-download-the-workflow-template-files)  
4. [Modify the variables in the Remove_Human_Reads.config file](#4-modify-the-variables-in-the-config-file)  
5. [Run the workflow](#5-run-the-workflow)  

<br>

---

### 1. Install conda, mamba, and `genelab-utils` package

> We recommend installing a Miniconda, Python3 version appropriate for your system, as exemplified in [the above link](https://astrobiomike.github.io/unix/conda-intro#getting-and-installing-conda).  

Reviewer: The text "the above link" isn't referring to anything above and should be updated to be more specific about where its pointing.
```suggestion
We recommend installing a Miniconda, Python3 version appropriate for your system, as exemplified in [the Happy Belly Bioinformatics conda tutorial](https://astrobiomike.github.io/unix/conda-intro#getting-and-installing-conda).  
```

Once conda is installed on your system, we recommend installing [mamba](https://github.com/mamba-org/mamba#mamba), as it generally allows for much faster conda installations:

```bash
conda install -n base -c conda-forge mamba
```

> > You can read a quick intro to mamba [here](https://astrobiomike.github.io/unix/conda-intro#bonus-mamba-no-5) if wanted.

Reviewer: ```suggestion
> You can read a quick intro to mamba [here](https://astrobiomike.github.io/unix/conda-intro#bonus-mamba-no-5) if desired.
```

We recommend installing Singularity on a system wide level as per the associated [documentation](https://docs.sylabs.io/guides/3.10/admin-guide/admin_quickstart.html).

> Note: Singularity is also available through [Anaconda](https://anaconda.org/conda-forge/singularity).

<br>

---

### 3. Download the workflow template files
All workflow files for removing human reads from metagenomics data are in the [workflow_code](workflow_code) directory. To get a copy of the latest NF_MGRemoveHumanReads-A version on to your system, run the following command:

```bash
GL-get-workflow NF_MGRemoveHumanReads-A
```

This downloaded the workflow into a directory called `NF_MGRemoveHumanReads-*/`, with the workflow version number at the end.

> Note: If wanting an earlier version, the wanted version can be provided as an optional argument like so:
> ```bash
> GL-get-workflow NF_MGRemoveHumanReads-A --wanted-version 1.0.0
> ```

<br>

---

### 4. Modify the variables in the Remove_Human_Reads.config file

> Once you've downloaded the workflow template, you can modify the variables in the [Remove_Human_Reads.config](workflow_code/Remove_Human_Reads.config) file as needed. For example, you will have to provide a text file containing a single-column list of unique sample identifiers (see an example of how to set this up below - if you are running the example dataset, this file is provided in the [workflow_code](workflow_code) directory [here](workflow_code/unique-sample-IDs.txt)). You will also need to indicate the path to your input data (raw reads) and the root directory for where the kraken2 reference database should be stored (it will be setup automatically). Additionally, if necessary, you'll need to modify each variable in the [Remove_Human_Reads.config](workflow_code/Remove_Human_Reads.config) file to be consistent with the study you want to process and the machine you're using. 

Reviewer: The file in the repository is named with a lowercase "reads.config" instead of uppercase "Reads.config". Please be consistent.

Please be super specific about which values to update in the config file. For example, the "$projectDir" variable that is used throughout that file is not defined anywhere and this will confuse users who may assume that $projectDir is somehow automatically set by nextflow (it is not).

> Note: If you are unfamiliar with how to specify paths, one place you can learn more is [here](https://astrobiomike.github.io/unix/getting-started#the-unix-file-system-structure).  

**Example for how to create a single-column list of unique sample identifiers from your raw data file names**

For example, if you only want to process a subset of the read files within the reads directory and have paired-end read data for 2 samples located in `../Raw_Sequence_Data/` relative to your workflow directory, that would look like this:

```bash
ls ../Raw_Sequence_Data/
```

```
Sample-1_R1.fastq.gz
Sample-1_R2.fastq.gz
Sample-2_R1.fastq.gz
Sample-2_R2.fastq.gz
```

You would set up your `unique-sample-IDs.txt` file as follows:

```bash
cat unique-sample-IDs.txt
```

```
Sample-1
Sample-2
```

<br>

---

### 5. Run the workflow

While in the directory holding the NextFlow file, .config file, and other workflow files that you downloaded in [step 3](#3-download-the-workflow-template-files), here is one example command of how to run the workflow:

```bash

> nextflow run *path/to/Remove_Human_Reads.nf* -ansi-log false -specify_reads false

Reviewer: This command does not work because it assumes that the config file is named "nextflow.config". It is OK to have the config named "Remove_Human_Reads.config", but then you have to specify that with "nextflow -c Remove_Human_Reads.config run ...." in order for nextflow to actually use it.

> nextflow run *path/to/Remove_Human_Reads.nf* -ansi-log false -specify_reads false

Reviewer: All nextflow command-line parameters should be specified with "--" and not "-". Instead of using a dummy path ("path/to/Remove_Human_reads.nf"), it is better to tell people to run the workflow from within the workflow_code folder or to provide the path to the workflow_code folder from the folder they downloaded the workflow_code to. I'm also not sure why you're telling them to turn the ansi-log off. It's better for the default behavior to have extra logging output.
```suggestion
nextflow run Remove_Human_Reads.nf --specify_reads false
```

```

> * `-ansi-log false` – specifies to print out each command being run to the screen

Reviewer: ```suggestion
* `--ansi-log false` – specifies to print out each command being run to the screen
```

> * `-resume` – continues to run the workflow using cached data from the previous run

Reviewer: ```suggestion
* `--resume` – continues to run the workflow using cached data from the previous run
```

> * `-specify_reads false` - processes all reads in the working directory, without requiring a sample ID list

Reviewer: ```suggestion
* `--specify_reads false` - processes all reads in the working directory, without requiring a sample ID list
```

See `nextflow -h` and [NextFlow's documentation](https://www.nextflow.io/docs/master/index.html) for more options and details.

A quick example can be run with the files included in the [workflow_code](workflow_code) directory after specifying a location for the reference database in the [Remove_Human_Reads.config](workflow_code/Remove_Human_Reads.config) file.

---

## Reference database info

> The database we use was built with kraken2 v2.1.1 as detailed below, and can be downloaded to run with this workflow (it's ~4.3 GB uncompressed).

Reviewer: Adding the link to the reference DB info page from the snakemake workflow README 
```suggestion
The database we use was built with kraken2 v2.1.1 as detailed below, and can be downloaded to run with this workflow (it's ~4.3 GB uncompressed). The steps for building it are described on the [reference database info page](https://github.com/nasa/GeneLab_Data_Processing/blob/master/Metagenomics/Remove_human_reads_from_raw_data/Workflow_Documentation/SW_MGRemoveHumanReads-A/reference-database-info.md).
```
